ipa-core/src/protocol/ipa_prf/malicious_security/quadratic_proofs.rs

```diff
@@ -0,0 +1,933 @@
+/// it is also the degree of polynomials p and q
+///
+/// the output on the right needs to be sent to the party on the left
+/// it is therefore the "right share" of the party on the left
+///
+/// the left part of the output is the "left share" of the proof generated by the party on the right
+/// this part is generated using PRSS and does not need to be sent,
+/// it remains with the party and is the input `proof_left` to the proof verification
+/// the outputs includes the two masks for the proofs, i.e. `mask_p`, `mask_q`
+///
+/// This function uses `ctx` for generating random values using PRSS
+/// and sending the right part of the proof to the left
+///
+/// # Errors
+/// propagates errors from send and receive
+///
+/// # Panics
+/// Does not panic
+pub async fn generate_proof<C, F>(
+    ctx: C,
+    u: &mut Vec<F>,
+    v: &mut Vec<F>,
+    r_f: usize,
+) -> Result<(NIDZKP<F>, NIDZKP<F>), Error>
+where
+    C: Context,
+    F: Field + Invert,
+{
+    // check length
+    debug_assert_eq!(u.len(), v.len());
+    // check that field is large enough to hold evaluation points
+    debug_assert!(F::BITS > usize::BITS - r_f.leading_zeros() + 1);
+    // record, used as counter for PRSS
+    let mut record = 0usize;
+
+    // proof length
+    #[allow(
+        clippy::cast_possible_truncation,
+        clippy::cast_sign_loss,
+        clippy::cast_precision_loss
+    )]
+    let proof_length = (u.len() as f64).log(r_f as f64).ceil() as usize + 1usize;
+    // get masks from PRSS
+    let mask_p = ctx.prss().generate_fields(RecordId::from(record)); //(F::ONE, F::ONE);
+    record += 1;
+    let mask_q = ctx.prss().generate_fields(RecordId::from(record));
+    // proof from the prover on the left, i.e. generated via `PRSS.left` rather than receiving it from the left
+    let mut output_left = NIDZKP {
+        proofs: Vec::with_capacity(proof_length),
+        // set mask p(0) from `PRSS.left`
+        mask_p: mask_p.0,
+        // set mask q(0) from `PRSS.left`
+        mask_q: mask_q.0,
+    };
+    // right part of this helpers own proof, to be sent to the left
+    let mut own_proof_right = NIDZKP {
+        proofs: Vec::with_capacity(proof_length),
+        mask_p: F::ZERO,
+        mask_q: F::ZERO,
+    };
+
+    // precomputation for Lagrange
+    let mut e_p = vec![F::ZERO; 2 * r_f];
+    generate_evaluation_points(&mut e_p);
+    // precomputation for interpolation of `u`, `v`, i.e. degree `recursion_factor-1` polynomial
+    let mut ds = vec![F::ONE; r_f];
+    compute_lagrange_denominator(&e_p[0..r_f], &mut ds);
+    let mut base = vec![vec![F::ONE; r_f]; r_f];
+    compute_lagrange_base(&e_p[r_f..2 * r_f], &e_p[0..r_f], &ds, &mut base);
+
+    // iterate over recursions
+    while u.len() >= r_f {
+        // fill u and v with zeros such that it is a multiple of the recursion factor
+        let coset = u.len() % r_f;
+        if coset != 0 {
+            u.extend(std::iter::repeat(F::ZERO).take(coset));
```

Comment:
An alternative approach, when `u.len() % r_f != 0` is to adjust `r_f` for that iteration.

Yes, this incurs the additional cost of precomputing additional lagrange coefficients, but the authors even used different values per step. They say:

```
Since the first round computation has been significantly improved, our protocol uses the compression parameter λ = 32 in the first round and λ′ = 8 for the rest of rounds
```

In my spreadsheet, I used λ = 4 for all the rounds until the final round when I used λ = 2, since that's all I needed (given the length of the input vector).